File: app/src/main/java/com/sillytavern/android/data/remote/ApiService.kt
package com.sillytavern.android.data.remote

import com.sillytavern.android.data.model.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.sse.EventSource
import okhttp3.sse.EventSourceListener
import okhttp3.sse.EventSources
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*

interface OpenAIApiService {
    @POST("v1/chat/completions")
    @Headers("Content-Type: application/json")
    suspend fun chatCompletion(@Body request: OpenAIRequest): OpenAIResponse
    
    @Streaming
    @POST("v1/chat/completions")
    @Headers("Content-Type: application/json")
    fun streamChatCompletion(@Body request: OpenAIRequest): OpenAIResponse
}

interface ClaudeApiService {
    companion object {
        fun create(client: OkHttpClient): ClaudeApiService {
            return Retrofit.Builder()
                .client(client)
                .baseUrl("https://api.anthropic.com/")
                .addConverterFactory(GsonConverterFactory.create())
                .build()
                .create(ClaudeApiService::class.java)
        }
    }
    
    @POST("v1/messages")
    suspend fun createMessage(
        @Header("x-api-key") apiKey: String,
        @Header("anthropic-version") version: String = "2023-06-01",
        @Body request: Map<String, Any>
    ): Map<String, Any>
}

interface KoboldApiService {
    companion object {
        fun create(client: OkHttpClient): KoboldApiService {
            return Retrofit.Builder()
                .client(client)
                .baseUrl("http://127.0.0.1:5000/")
                .addConverterFactory(GsonConverterFactory.create())
                .build()
                .create(KoboldApiService::class.java)
        }
    }
    
    @POST("api/v1/generate")
    suspend fun generate(@Body request: KoboldRequest): KoboldResponse
    
    @GET("api/v1/model")
    suspend fun getModel(): Map<String, String>
    
    @GET("api/v1/info/version")
    suspend fun getVersion(): Map<String, String>
}

class ApiService(
    private val openAIApiService: OpenAIApiService,
    private val claudeApiService: ClaudeApiService,
    private val koboldApiService: KoboldApiService
) {
    suspend fun openAIChatCompletion(
        apiKey: String,
        request: OpenAIRequest
    ): Result<OpenAIResponse> {
        return try {
            Result.success(openAIApiService.chatCompletion(request))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun claudeCreateMessage(
        apiKey: String,
        request: Map<String, Any>
    ): Result<Map<String, Any>> {
        return try {
            Result.success(claudeApiService.createMessage(apiKey, request = request))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun koboldGenerate(
        baseUrl: String,
        request: KoboldRequest
    ): Result<KoboldResponse> {
        return try {
            Result.success(koboldApiService.generate(request))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    fun streamOpenAIChat(
        client: OkHttpClient,
        url: String,
        apiKey: String,
        request: OpenAIRequest,
        onToken: (String) -> Unit,
        onComplete: () -> Unit,
        onError: (Throwable) -> Unit
    ): EventSource {
        val eventSourceFactory = EventSources.createFactory(client)
        val requestBody = com.google.gson.Gson().toJson(request)
            .toRequestBody("application/json".toMediaType())
        
        val httpRequest = Request.Builder()
            .url(url)
            .header("Authorization", "Bearer $apiKey")
            .header("Content-Type", "application/json")
            .post(requestBody)
            .build()
        
        val listener = object : EventSourceListener() {
            private val buffer = StringBuilder()
            
            override fun onEvent(
                eventSource: EventSource,
                id: String?,
                type: String?,
                data: String
            ) {
                if (data == "[DONE]") {
                    onComplete()
                    return
                }
                
                try {
                    val response = com.google.gson.Gson().fromJson(data, OpenAIResponse::class.java)
                    response.choices.firstOrNull()?.delta?.content?.let { content ->
                        buffer.append(content)
                        onToken(content)
                    }
                } catch (e: Exception) {
                }
            }
            
            override fun onClosed(eventSource: EventSource) {
                if (buffer.isNotEmpty()) {
                    onComplete()
                }
            }
            
            override fun onFailure(eventSource: EventSource, t: Throwable?, response: okhttp3.Response?) {
                onError(t ?: Exception("Stream failed: ${response?.message}"))
            }
        }
        
        return eventSourceFactory.newEventSource(httpRequest, listener)
    }
}

File: lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class ApiService {
  String _baseUrl = 'http://localhost:8000';
  String _apiKey = '';
  String _apiType = 'openai';
  String _model = 'gpt-3.5-turbo';

  void updateSettings({
    String? baseUrl,
    String? apiKey,
    String? apiType,
    String? model,
  }) {
    if (baseUrl != null) _baseUrl = baseUrl;
    if (apiKey != null) _apiKey = apiKey;
    if (apiType != null) _apiType = apiType;
    if (model != null) _model = model;
  }

  Future<String> sendMessage(List<Map<String, String>> messages) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/v1/chat/completions'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $_apiKey',
        },
        body: jsonEncode({
          'model': _model,
          'messages': messages,
          'stream': false,
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['choices'][0]['message']['content'];
      } else {
        throw Exception('API Error: ${response.statusCode} - ${response.body}');
      }
    } catch (e) {
      throw Exception('Connection failed: $e');
    }
  }

  Stream<String> sendMessageStream(List<Map<String, String>> messages) async* {
    try {
      final request = http.Request(
        'POST',
        Uri.parse('$_baseUrl/v1/chat/completions'),
      );
      request.headers['Content-Type'] = 'application/json';
      request.headers['Authorization'] = 'Bearer $_apiKey';
      request.body = jsonEncode({
        'model': _model,
        'messages': messages,
        'stream': true,
      });

      final streamedResponse = await http.Client().send(request);
      
      await for (final chunk in streamedResponse.stream.transform(utf8.decoder).transform(const LineSplitter())) {
        if (chunk.startsWith('data: ')) {
          final data = chunk.substring(6);
          if (data == '[DONE]') break;
          try {
            final json = jsonDecode(data);
            final content = json['choices'][0]['delta']['content'];
            if (content != null) {
              yield content;
            }
          } catch (_) {}
        }
      }
    } catch (e) {
      yield* Stream.error(Exception('Stream failed: $e'));
    }
  }

  Future<Map<String, dynamic>> getModels() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/v1/models'),
        headers: {
          'Authorization': 'Bearer $_apiKey',
        },
      );
      
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return {};
    } catch (e) {
      return {};
    }
  }
}
